zio-config/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -1,13 +1,66 @@
 package zio.config
 
-import zio.IO
+import zio.config.ReadErrors.ReadError
+import zio.system.System
+import zio.{ IO, ZIO }
 
-trait ConfigSource {
-  val configService: ConfigSource.Service
+trait ConfigSource[K, V] {
+  val configSourceService: ConfigSource.Service[K, V]
 }
 
 object ConfigSource {
-  trait Service {
-    def getConfigValue(path: String): IO[ReadError, String]
+  trait Service[K, V] {
+    def getConfigValue(path: List[K]): IO[ReadError[K, V], V]
   }
+
+  val fromEnv: ZIO[System, Nothing, ConfigSource[String, String]] =
+    ZIO.access { env =>
+      new ConfigSource[String, String] {
+        val configSourceService: ConfigSource.Service[String, String] =
+          (path: List[String]) => {
+            val key = path.mkString("_")
+            env.system
+              .env(key)
+              .mapError { err =>
+                ReadError.FatalError(key, err): ReadError[String, String]
+              }
+              .flatMap(IO.fromOption(_))
+              .mapError(_ => ReadError.MissingValue(key))
+          }
+      }
+    }
+
+  val fromProperty: ZIO[System, Nothing, ConfigSource[String, String]] =
+    ZIO.access { env =>
+      new ConfigSource[String, String] {
+        val configSourceService: ConfigSource.Service[String, String] =
+          (path: List[String]) => {
+            val key = path.mkString(".")
+            env.system
+              .property(key)
+              .mapError { err =>
+                ReadError.FatalError(key, err): ReadError[String, String]
+              }
+              .flatMap(IO.fromOption(_))
+              .mapError(_ => ReadError.MissingValue(key))
+          }
+      }
+    }
+
+  def fromMap(map: Map[String, String], delimiter: String = "."): ConfigSource[String, String] =
+    new ConfigSource[String, String] {
+      val configSourceService: ConfigSource.Service[String, String] =
+        (path: List[String]) => {
+          val key = path.mkString(delimiter)
+          ZIO.fromOption(map.get(key)).mapError { _ =>
+            ReadError.MissingValue(key): ReadError[String, String]
+          }
+        }
+    }
+
+  def fromPropertyTree(
```

Comment:
Maybe: `fromPropertyTreeString`? At the risk of making method names too long. 😆 

In theory, we could have `fromPropertyTree[K, V]` at some point.